pkg/storage/swift/swift.go

```diff
@@ -1,17 +1,76 @@
 package swift
 
 import (
+	"reflect"
+
 	corev1 "k8s.io/api/core/v1"
 
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack"
+	"github.com/gophercloud/gophercloud/openstack/objectstorage/v1/containers"
+
+	operatorapi "github.com/openshift/api/operator/v1"
+
 	imageregistryv1 "github.com/openshift/cluster-image-registry-operator/pkg/apis/imageregistry/v1"
 	regopclient "github.com/openshift/cluster-image-registry-operator/pkg/client"
+	"github.com/openshift/cluster-image-registry-operator/pkg/storage/util"
 )
 
 type driver struct {
 	Config  *imageregistryv1.ImageRegistryConfigStorageSwift
 	Listers *regopclient.Listers
 }
 
+var (
+	swiftClient *gophercloud.ServiceClient
+)
+
+// getSwiftClient returns a client that allows us to interact with the OpenStack Swift service
+func (d *driver) getSwiftClient(cr *imageregistryv1.Config) (*gophercloud.ServiceClient, error) {
+	if swiftClient != nil {
+		return swiftClient, nil
+	}
+
+	// First read config from the secret with clouds.yaml
+	if cr.Spec.Storage.Swift.CloudsSecret == nil || cr.Spec.Storage.Swift.CloudName == "" {
+		err := setConfigFromSecret(cr.Spec.Storage.Swift)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// Try to read credentials from the native secret
+		err := setSwiftCredentials(d.Listers, cr.Spec.Storage.Swift)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	opts := gophercloud.AuthOptions{
```

Comment:
I'm not a huge fan of building `AuthOpts` like this because clouds have different credentials. Some of them use regions, others don't. Others use trusts, etc. If you have the `clouds` object, you can tell `gophrecloud` to build the opts object itself.

Take a look here:

https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/master/pkg/cloud/openstack/clients/machineservice.go#L91

https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/master/pkg/cloud/openstack/clients/machineservice.go#L146

Reply:
Okay, makes sense.